Source/Core/VideoCommon/VertexShaderGen.cpp

```diff
@@ -191,6 +191,35 @@ ShaderCode GenerateVertexShaderCode(APIType api_type, const ShaderHostConfig& ho
 
   out.WriteFmt("VS_OUTPUT o;\n");
 
+  // xfmem.numColorChans controls the number of color channels available to TEV, but we still need
+  // to generate all channels here, as it can be used in texgen. Cel-damage is an example of this.
+  out.WriteFmt("float4 vertex_color_0, vertex_color_1;\n");
+
+  // To use color 1, the vertex descriptor must have color 0 and 1.
+  // If color 1 is present but not color 0, it is used for lighting channel 0.
+  const bool use_color_1 =
+      (uid_data->components & (VB_HAS_COL0 | VB_HAS_COL1)) == (VB_HAS_COL0 | VB_HAS_COL1);
+  for (u32 color = 0; color < NUM_XF_COLOR_CHANNELS; color++)
+  {
+    if ((color == 0 || use_color_1) && uid_data->components & (VB_HAS_COL0 << color))
+    {
+      // Use color0 for channel 0, and color1 for channel 1 if both colors 0 and 1 are present.
+      out.WriteFmt("vertex_color_%u = rawcolor%u;\n", color, color);
+    }
+    else if (color == 0 && (uid_data->components & VB_HAS_COL1))
```

Comment:
```suggestion
    else if (color == 0 && (uid_data->components & VB_HAS_COL1) != 0)
```

To be explicit about the comparison